src/libcore/char.rs

```diff
@@ -176,6 +172,51 @@ pub unsafe fn from_u32_unchecked(i: u32) -> char {
     transmute(i)
 }
 
+#[stable(feature = "char_convert", since = "1.12.0")]
+impl From<char> for u32 {
+    #[inline]
+    fn from(c: char) -> Self {
+        c as u32
+    }
+}
+
+#[stable(feature = "char_convert", since = "1.12.0")]
+impl From<u8> for char {
+    #[inline]
+    fn from(i: u8) -> Self {
+        i as char
+    }
+}
+
+#[unstable(feature = "try_from", issue = "33417")]
+impl TryFrom<u32> for char {
+    type Err = CharTryFromError;
+
+    #[inline]
+    fn try_from(i: u32) -> Result<Self, Self::Err> {
+        if i > MAX as u32 {
+            Err(CharTryFromError::OutOfRange)
+        } else if i >= 0xD800 && i <= 0xDFFF {
+            Err(CharTryFromError::SurrogateCodePoint)
+        } else {
+            Ok(unsafe { from_u32_unchecked(i) })
+        }
+    }
+}
+
+/// The error type returned when a conversion from u32 to char fails.
+#[unstable(feature = "try_from", issue = "33417")]
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub enum CharTryFromError {
+    /// The given integer is beyond Unicode’s last code point, 0x10FFFF
+    OutOfRange,
+
+    /// The given integer is in the range of surrogate code points, 0xD800 to 0xDFFF.
+    ///
+    /// Surrogates are used in the UTF-16 encoding, and therefore are not characters.
+    SurrogateCodePoint,
+}
```

Comment:
We're never going to need to add any extra cases to this enum, right? Should we stick a `__ForExtensibility` variant just in case?